src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java

```diff
@@ -1,35 +1,589 @@
+  /**
+   * text and confidence (and any additional meta data) to be published
+   */
+  public static class RecognizedResult {
+    public Double confidence;
+    public boolean isFinal = false;
+    public String text;
+  }
+
+  /**
+   * generalized list of languages and their codes - if useful
+   */
+  static protected Map<String, String> languages = new HashMap<>();
 
   private static final long serialVersionUID = 1L;
 
-  String wakeWord = null;
+  /**
+   * all currently attached services
+   */
+  protected Set<String> attached = new TreeSet<>();
+
+  protected HashMap<String, Message> commands = new HashMap<>();
+
+  /**
+   * status of listening
+   */
+  protected boolean isListening = false;
+
+  /**
+   * status when wake word is used and is ready to publish recognized events
+   */
+  protected boolean isAwake = false;
+
+  /**
+   * status of publishing recognized text
+   */
+  protected boolean isRecognizing = false;
+
+  /**
+   * current language code for recognition
+   */
+  protected String language = "en-US";
+
+  @Deprecated /* remove ! - is from webkit - should be handled in js */
+  protected long lastAutoListenEvent = System.currentTimeMillis();
+
+  protected String lastThingRecognized = null;
+
+  /**
+   * prevents execution of command unless a lock out phrase is recognized
+   */
+  @Deprecated /* use wake word */
+  protected boolean lockOutAllGrammar = false;
+
+  /**
+   * phrase to unlock commands such as "power up"
+   */
+  @Deprecated /* use wake word */
+  protected String lockPhrase = "";
+
+  protected boolean lowerCase = true;
+
+  /**
+   * Used in conjunction with a SpeechSythesis
+   */
+  // FIXME - probably not needed - what needs to happen is SpeechSynthesis
+  // should "only" affect recognizing NOT listening
+  // and an end user affect listening ...
+  // protected boolean speaking = false;
+
+  /**
+   * Wake word functionality is activated when it is set (ie not null) This
+   * means recognizing events will be processed "after" it hears the wake word.
+   * It will continue to publish events until a idle timeout period is reached.
+   * It can continue to listen after this, but it will not publish. It fact, it
+   * 'must' keep listening since in this idle state it needs to search for the
+   * wake word
+   */
+  protected String wakeWord = null;
+
+  /**
+   * number of seconds of silence after the initial wake word is used that it
+   * the wake word will be needed to activate again null == unlimited
+   */
+  protected Integer wakeWordIdleTimeoutSeconds = 10;
+
+  protected Long lastWakeWordTs = null;
+
+  protected boolean isSpeaking = false;
+
+  /**
+   * wait for 1 sec after my speaking has ended
+   */
+  protected long afterSpeakingPauseMs = 1000;
 
   public AbstractSpeechRecognizer(String n, String id) {
     super(n, id);
+    languages.put("en-US", "English - United States");

```

Comment:
This list of langauges is specific to the WebKitSpeechRecognizer class..  it should not be in the base abstract class.

Reply:
this is changed locally - my local has SpeechRecognizer and SpeechSynthesis both extend LocaleProvider interface - and the respective abstracts contain "default" sets of Locales, if a service needs to implement some limited subset, it simply overrides the abstract and returns the set it actually supports in a "getLocales()" call

The nice part is we're now always dealing with the same format and same methods from the interface.